### PR TITLE
Add zoom to the camera

### PR DESCRIPTION
**Pull request by luanpotter**
# Description

Add option to zoom in (or out) the camera. Also adding docs for Camera and Viewport as there weren't before.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `./scripts/format.sh` and the Flame analyzer (`./scripts/analyze.sh`) does not report any problems.
- [x] I read and followed the [Flutter Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

### CODE REVIEW

File: doc/camera_and_viewport.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# Camera and Viewport
+
+When rendering on Flutter, the regular coordinate space used are logical pixels. That means one pixel for Flutter is already not necessarily one real pixel on the device, because of the [device's pixel ratio](https://api.flutter.dev/flutter/widgets/MediaQueryData/devicePixelRatio.html). When it gets to the Flame level, we always consider the most fundamental level to be logical pixels, so all the device specific complexity is abstracted away.
```
Comment (spydon) [inline doc/camera_and_viewport.md:3]: Max line length of 100 to be in line with the other docs

File: packages/flame/lib/src/game/camera.dart [hunk 2]
```diff
@@ -142,8 +143,33 @@ class Camera {
   /// add any non-smooth movement.
   Rect? worldBounds;
 
+  /// If set, the camera will zoom by this ratio. This be greater than 1 (in)
+  /// or smaller (out) but should be always greater than zero.
```
Comment (spydon) [inline packages/flame/lib/src/game/camera.dart:147]:
```suggestion
  /// or smaller (out) but should always be greater than zero.
```

Comment (spydon) [reply packages/flame/lib/src/game/camera.dart:147]: This sentence is a bit confusing over all, less than 1 should be zoom out right?

File: doc/camera_and_viewport.md (hunk 1 above)
Comment (luanpotter) [reply doc/camera_and_viewport.md:3]: really? didn't know about this rule, is it enforced by our linters or written anywhere?

File: packages/flame/lib/src/game/camera.dart (hunk 2 above)
Comment (luanpotter) [reply packages/flame/lib/src/game/camera.dart:147]: yeah I forgot a `can` as well

File: doc/camera_and_viewport.md (hunk 1 above)
Comment (spydon) [reply doc/camera_and_viewport.md:3]: It's not, we don't have an md linter, me and @wolfenrain talked about introducing one though

File: packages/flame/test/game/camera_and_viewport_test.dart [hunk 3]
```diff
@@ -304,6 +304,28 @@ void main() {
...
+      game.camera.zoom = 2;
+
+      final p = TestComponent(Vector2.all(100.0))..anchor = Anchor.center;
+      game.add(p);
+      game.update(0);
+
+      final canvas = MockCanvas();
+      game.render(canvas);
+      expect(
+        canvas.methodCalls
```
Comment (spydon) [inline packages/flame/test/game/camera_and_viewport_test.dart:319]: This is great, didn't know about that method

**Review by spydon — APPROVED** [review]
All looks good! Except the line length on the doc

File: doc/camera_and_viewport.md (hunk 1 above)
Comment (wolfenrain) [reply doc/camera_and_viewport.md:3]: Yeah I have been working on enforcing it globally, hope to have something for it soon :-)

File: doc/camera_and_viewport.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+The Viewport is an attempt to unify multiple screen (or, rather, game widget) sizes into a single configuration for your game by translating and resizing the canvas.
+
+The `Viewport` interface has multiple implementations and can be used from scratch on your `Game` or, if you are using `BaseGame` instead, it's already built-in (with a default no-op viewport).
+
+These are the viewports available to pick from (or you can implement the interface yourself to suit your needs):
+
+ * `DefaultViewport`: this is the no-op viewport that is associated by default with any `BaseGame`.
+ * `FixedResolutionViewport`: this viewport transforms your Canvas so that, from the game perspective, the dimensions are always set to a fixed pre-defined value. This means it will scale the game as much as possible and add black bars if needed.
+ * More viewports are planned!
```
Comment (erickzanardo) [inline doc/camera_and_viewport.md:17]: Not sure we need this line her

```suggestion
```

File: doc/camera_and_viewport.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ * follow: you can use this method so that your camera continuously "follow" an object (for example, a `PositionComponent`). This is not smooth because the movement of the followed object itself is assumed to already be smooth (i.e. if your character teleport the camera will also immediately teleport).
+ * world bounds: when using follow, you can optionally define the bounds of the world. If that is done, the camera will stop following/moving so that out-of-bounds areas are not shown (as long as the world is bigger than the screen).
+
+Finally the second transformation that the camera applies is scaling. That allows for dynamic zooming, and it's controlled by the `zoom` field. There is no zoom speed, that must be controlled by you when changing. The `zoom` variable is immediately applied.
+
+When dealing with events, it is imperative to convert screen coordinates to world coordinates (or, for some reasons, you might want to do the reverse). The Camera provides two functions, `screenToWorld` and `worldToScreen` to easily convert between these coordinate spaces.
```
Comment (erickzanardo) [inline doc/camera_and_viewport.md:45]:
```suggestion
When dealing with touch/click events, it is imperative to convert screen coordinates to world coordinates (or, for some reasons, you might want to do the reverse). The Camera provides two functions, `screenToWorld` and `worldToScreen` to easily convert between these coordinate spaces.
```

**Review by erickzanardo — APPROVED** [review]
LGTM great improve on the docs

File: doc/camera_and_viewport.md (hunk 1 above)
Comment (luanpotter) [reply doc/camera_and_viewport.md:3]: is there an easy way to format for now or do I have to manually split the lines?

File: doc/camera_and_viewport.md (hunk 5 above)
Comment (luanpotter) [reply doc/camera_and_viewport.md:45]: not only touch or click, any events. drag, scroll, etc

File: doc/camera_and_viewport.md (hunk 4 above)
Comment (luanpotter) [reply doc/camera_and_viewport.md:17]: I found it weird to have a list with only two, but let me remove

File: doc/camera_and_viewport.md (hunk 5 above)
Comment (erickzanardo) [reply doc/camera_and_viewport.md:45]: Drag are touch events right? Scroll you don't need to translate I think. Anyway the idea here is to make it clear that those events are input events, I think it wasn't very clear which events the sentence is talking about

Comment (luanpotter) [reply doc/camera_and_viewport.md:45]: scrolls also have positions. I will add input

File: doc/camera_and_viewport.md (hunk 1 above)
Comment (luanpotter) [reply doc/camera_and_viewport.md:3]: figured it out!

**Pull request merged**